api/v1/relay/relay.proto

```diff
@@ -0,0 +1,154 @@
+// limitations under the License.
+
+syntax = "proto3";
+
+import "google/protobuf/timestamp.proto";
+import public "flow/flow.proto";
+
+package relay;
+
+service HubbleRelay {
+    rpc GetEvents(GetEventsRequest) returns (stream GetEventsResponse) {}
+}
+
+// Here I didn't include "follow", "until", and "number". This request assumes follow,
+// and lets the client decide when to end the request, whether it's based on timestamp
+// or the number of responses received.
+message GetEventsRequest{
+    repeated EventFilter blacklist = 1;
+    repeated EventFilter whitelist = 2;
+    google.protobuf.Timestamp since = 3;
+}
+
+message GetEventsResponse{
+    string node = 1;
+    google.protobuf.Timestamp timestamp = 2;
+    oneof event {
+        flow.Flow flow = 3;
+        K8sNamespaceAdded k8s_namespace_added = 4;
+        K8sNamespaceDeleted k8s_namespace_deleted = 5;
+        ServiceAdded service_added = 6;
+        ServiceAdded service_deleted = 7;
+        // This is needed in case the relay server knows that the service exists,
+        // but it doesn't know when it got created. I didn't add the exists event
+        // for k8s namespace because the creation timestamp is always known.
+        ServiceExists service_exists = 8;
+        // It's unclear how we can define what it means for service links to get
+        // added/deleted. For now I'm deferring the decision for whether a particular
+        // link should be included in the service map to the client. Maybe the server
+        // can periodically send service_link events for active links to keep-alive,
+        // and the client can expire older links based on their timestamp.
+        ServiceLink service_link_exists = 9;
+    }
+}
+
+// Here I added different filter types for different events. Another option is to reuse
+// and extend FlowFilter to filter all the events, but I worry that it might become too
+// confusing in terms of figuring out which filter fields apply to which events.
+message EventFilter {
+    repeated EventType type = 1;
+    repeated flow.FlowFilter flow_filter = 2;
+    repeated ServiceFilter service_filter = 3;
+    repeated ServiceLinkFilter service_link_filter = 4;
+}
+
+enum EventType {
+    UNKNOWN_EVENT = 0;
+    FLOW = 1;
+    K8S_NAMESPACE_ADDED = 2;
+    K8S_NAMESPACE_DELETED = 3;
+    SERVICE_ADDED = 4;
+    SERVICE_DELETED = 5;
+    SERVICE_EXISTS = 6;
+    SERVICE_LINK_EXISTS = 7;
+}
+
+message K8sNamespace {
+    string id = 1;
+    string name = 2;
+    google.protobuf.Timestamp creation_timestamp = 3;
+}
+
+message K8sNamespaceAdded {
+    K8sNamespace namespace = 1;
+}
+
+message K8sNamespaceDeleted {
+    K8sNamespace namespace = 1;
+}
+
+// IP protocols. The values of enums do not correspond to actual IP protocol numbers.
+enum IPProtocol {
+    UNKNOWN_IP_PROTOCOL = 0;
+    TCP = 1;
+    UDP = 2;
+    ICMP_V4 = 3;
+    ICMP_V6 = 4;
+}
+
+// Come up with a better name for Service.
+message Service {
```

Comment:
will eventually need enforcement status, as per initial discussion

Reply:
yes. i excluded it for the time being. we can introduce it with policy-related events.

Reply:
on the second thought why wait. added these fields.